src/sentry/tasks/integrations/migrate_alert_rules.py

```diff
@@ -0,0 +1,109 @@
+        organization_id=organization_id, integration_id=integration_id
+    )
+    if not integration:
+        raise Integration.DoesNotExist
+    if not organization_integration:
+        raise OrganizationIntegration.DoesNotExist
+
+    seen_keys = {team["integration_key"] for team in organization_integration.config["team_table"]}
+
+    for project in Project.objects.filter(organization_id=organization_id):
+        plugin = None
+        for p in plugins.for_project(project):
+            if isinstance(p, OpsGeniePlugin) and p.is_configured(project):
+                plugin = p
+                break
+        if not plugin:
+            continue
+
+        # migrate key
+        api_key = plugin.get_option("api_key", project)
+        if api_key not in seen_keys:
+            seen_keys.add(api_key)
+            config = organization_integration.config
+            team_table = config["team_table"]
+            team = {
+                "team": project.name + " [MIGRATED]",
+                "id": str(organization_id) + "-" + project.name,
+                "integration_key": api_key,
+            }
+            team_table.append(team)
+            config.update({"team_table": team_table})
+            with transaction.atomic(router.db_for_write(OrganizationIntegration)):
+                integration_service.update_organization_integration(
+                    org_integration_id=organization_integration.id, config=config
+                )
+                logger.info(
+                    "api_key.migrated",
+                    extra={
+                        "integration_id": integration_id,
+                        "organization_id": organization_id,
+                        "project_id": project.id,
+                        "team_name": team["team"],
+                        "plugin": plugin.slug,
+                    },
+                )
+        else:
+            team = [team for team in team_table if team["integration_key"] == api_key][0]
+            logger.info(
+                "api_key.key_already_exists",
+                extra={
+                    "integration_id": integration_id,
+                    "organization_id": organization_id,
+                    "project_id": project.id,
+                    "team_name": team["team"],
+                    "plugin": plugin.slug,
+                },
+            )
+
+        # migrate alert rules
+        for rule in Rule.objects.filter(project_id=project.id):
+            actions = rule.data["actions"]
+            if ALERT_LEGACY_INTEGRATIONS in actions:
+                actions.append(
+                    {
+                        "id": "sentry.integrations.opsgenie.notify_action.OpsgenieNotifyTeamAction",
+                        "account": organization_integration.id,
+                        "team": team["id"],
+                    }
+                )
+                with transaction.atomic(router.db_for_write(Rule)):
```

Comment:
You will still need a transaction for this rule update though. What should be done though is that every `rule.save()` should happen within the same context manager. This means it needs to be outside the rule loop, or you do two loops, one to get only rules with legacy alerts, and the second one wrapped in the transaction.

The reason for this is that we only want two possibilities: Either all alerts for a project are migrated, or none are (because one failed).